### DISCUSSION TITLE
GluonTS transformation issue

### DISCUSSION THREAD

**Discussion by miteshyadav** [awslabs/gluonts#1479] [Q&A]
## Description
In order to handle missing values in the dataset, I am using GluonTS's ObservedValueIndicator transformation function to flag missing and observed values. After applying the transformation on the dataset, the training stagnates at 0% and wouldn't progress. I followed the exact template given in the Extended Forecasting Tutorial in GluonTS and the error still persists. I wonder if there is any workaround to instead just create my own flag for missing values using feat_dynamic_cat field. Let me know your thoughts. Thanks

**Comment by lostella** [comment]
@miteshyadav can you include a runnable example of the code that you're running?

**Comment by miteshyadav** [comment]
Sure @lostella 

Example is exactly the same as here: https://gluon-ts.s3-accelerate.dualstack.amazonaws.com/master/tutorials/forecasting/extended_tutorial.html

The custom dataset is also exactlythe same as in the above example. I use the train_tf object generated after transformation to feed it into the estimator:
predictor = estimator.train(training_data=train_tf)

That is when the training halts at 0% forever and doesn't budge.

**Comment by lostella** [comment]
The `train_tf` object defined in cell 34

```
train_tf = transformation(iter(train_ds), is_train=True)
```

is a generator (see output of cell 35). This means that once it gets iterated once, it will be exhausted. So it is not an object that should be used for training purposes: there, one usually wants to iterate over the dataset multiple times to fit the model well.

To apply a transformation persistently you should use the `TransformedDataset` class, see here: https://gluon-ts.s3-accelerate.dualstack.amazonaws.com/master/api/gluonts/gluonts.transform.dataset.html#module-gluonts.transform.dataset

**Reply by mbrhd** [reply]
Hi @lostella 

Thanks for this reply. I'm getting an issue with the `TransformedDataset` class.

I've used the same transformation defined by the Extended tutorial and set this transformation as input of my`TransformedDataset'. 

My code below:

```
from gluonts.transform import (
    AddAgeFeature,
    AddObservedValuesIndicator,
    Chain,
    ExpectedNumInstanceSampler,
    InstanceSplitter,
    SetFieldIfNotPresent,
)

from gluonts.transform import TransformedDataset

def create_transformation(freq, context_length, prediction_length):
    return Chain(
        [
            AddObservedValuesIndicator(
                target_field=FieldName.TARGET,
                output_field=FieldName.OBSERVED_VALUES,
            ),
            AddAgeFeature(
                target_field=FieldName.TARGET,
                output_field=FieldName.FEAT_AGE,
                pred_length=prediction_length,
                log_scale=True,
            ),
            InstanceSplitter(
                target_field=FieldName.TARGET,
                is_pad_field=FieldName.IS_PAD,
                start_field=FieldName.START,
                forecast_start_field=FieldName.FORECAST_START,
                instance_sampler=ExpectedNumInstanceSampler(
                    num_instances=1,
                    min_future=prediction_length,
                ),
                past_length=context_length,
                future_length=prediction_length,
                time_series_fields=[
                    FieldName.FEAT_AGE,
                    FieldName.FEAT_DYNAMIC_REAL,
                    FieldName.OBSERVED_VALUES,
                ],
            ),
        ]
    )

transformation = create_transformation(
    freq='D', 
    context_length=context_length, 
    prediction_length=prediction_length
)

train_tf = TransformedDataset(train_ds, transformation, is_train=True)
test_tf = TransformedDataset(test_ds, transformation, is_train=False)

gluonts_datasets = TrainDatasets(
    metadata=MetaData(
                freq=freq,
                target={'name': 'quantity'},
                prediction_length = prediction_length
    ),
    train=train_tf,
    test=test_tf
)

reloaded_dataset = load_datasets(
                        metadata=os.path.join(local_path, "metadata"),
                        train=os.path.join(local_path, "train"),
                        test=os.path.join(local_path, "test")
                   )
```

After reloading my dataset and doing `next(iter(reloaded_dataset.train))` I'm getting the following error GluonTSDataError: Object is missing a required field `target`

Can you help me?
